File: packages/pvm-assemblyscript/assembly/instructions/shifts-64-register.ts
import {
  OPCODE_SHAR_R_64,
  OPCODE_SHLO_L_64,
  OPCODE_SHLO_R_64,
} from '../config'
import { InstructionContext, InstructionResult } from '../types'
import { BaseInstruction } from './base'

/**
 * SHLO_L_64 instruction (opcode 0xCF)
 * Gray Paper formula: reg'_D = (reg_A · 2^(reg_B mod 64)) mod 2^64
 */
export class SHLO_L_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_SHLO_L_64
  name: string = 'SHLO_L_64'

  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValueAs64(context.registers, registerA)
    const valueB = this.getRegisterValueAs64(context.registers, registerB)
    const shiftAmount = valueB % u64(64)
    const result = valueA << shiftAmount

    this.setRegisterValue(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

/**
 * SHLO_R_64 instruction (opcode 0xD0)
 * Gray Paper formula: reg'_D = floor(reg_A ÷ 2^(reg_B mod 64))
 */
export class SHLO_R_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_SHLO_R_64
  name: string = 'SHLO_R_64'

  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValueAs64(context.registers, registerA)
    const valueB = this.getRegisterValueAs64(context.registers, registerB)
    const shiftAmount = valueB % u64(64)
    const result = valueA >> shiftAmount

    this.setRegisterValueWith64BitResult(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

/**
 * SHAR_R_64 instruction (opcode 0xD1)
 * Gray Paper formula: reg'_D = unsigned{floor(signed(reg_A) ÷ 2^(reg_B mod 64))}
 */
export class SHAR_R_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_SHAR_R_64
  name: string = 'SHAR_R_64'

  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValueAs64(context.registers, registerA)
    const valueB = this.getRegisterValueAs64(context.registers, registerB)
    const shiftAmount = valueB % u64(64)

    // Convert to signed for arithmetic shift
    const signedValue = this.toSigned64(valueA)
    const shiftedValue = signedValue >> i64(shiftAmount)
    const result = u64(this.toUnsigned64(shiftedValue))

    this.setRegisterValueWith64BitResult(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

File: packages/pvm-rust/src/instructions/shifts_64_register.rs
//! Shift by register (64-bit): SHLO_L_64, SHLO_R_64, SHAR_R_64.

use crate::config::{OPCODE_SHAR_R_64, OPCODE_SHLO_L_64, OPCODE_SHLO_R_64};
use crate::instructions::base::{
    arithmetic_shift_right_64, parse_three_registers, InstructionHandler,
};
use crate::types::{InstructionContext, InstructionResult};

fn get_register(registers: &[u64; 13], index: u8) -> u64 {
    registers.get(index as usize).copied().unwrap_or(0)
}

fn set_register(registers: &mut [u64; 13], index: u8, value: u64) {
    if (index as usize) < 13 {
        registers[index as usize] = value;
    }
}

// SHLO_L_64: reg'_D = (reg_A << (reg_B mod 64)) mod 2^64
pub struct ShloL64Instruction;
impl ShloL64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for ShloL64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_SHLO_L_64)
    }
    fn name(&self) -> &'static str {
        "SHLO_L_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_three_registers(context.operands);
        let a = get_register(context.registers, p.register_a);
        let shift = get_register(context.registers, p.register_b) % 64;
        let result = (a << shift) & 0xffff_ffff_ffff_ffff;
        set_register(context.registers, p.register_d, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// SHLO_R_64: reg'_D = floor(reg_A / 2^(reg_B mod 64))
pub struct ShloR64Instruction;
impl ShloR64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for ShloR64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_SHLO_R_64)
    }
    fn name(&self) -> &'static str {
        "SHLO_R_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_three_registers(context.operands);
        let a = get_register(context.registers, p.register_a);
        let shift = get_register(context.registers, p.register_b) % 64;
        let result = a >> shift;
        set_register(context.registers, p.register_d, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// SHAR_R_64: reg'_D = unsigned{floor(signed(reg_A) / 2^(reg_B mod 64))}
pub struct SharR64Instruction;
impl SharR64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for SharR64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_SHAR_R_64)
    }
    fn name(&self) -> &'static str {
        "SHAR_R_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_three_registers(context.operands);
        let a = get_register(context.registers, p.register_a);
        let shift = (get_register(context.registers, p.register_b) % 64) as u32;
        let result = arithmetic_shift_right_64(a, shift);
        set_register(context.registers, p.register_d, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}
